tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1264 @@
+    if (attempt >= spec_.retries->max_retries) {
+      return MaybeAnnotateStatus(
+          status,
+          tensorstore::StrCat("All ", attempt, " retry attempts failed"),
+          absl::StatusCode::kAborted, loc);
+    }
+
+    // https://cloud.google.com/storage/docs/retry-strategy#exponential-backoff
+    s3_retries.Increment();
+    auto delay = internal::BackoffForAttempt(
+        attempt, spec_.retries->initial_delay, spec_.retries->max_delay,
+        /*jitter=*/std::min(absl::Seconds(1), spec_.retries->initial_delay));
+        ScheduleAt(absl::Now() + delay,
+               WithExecutor(executor(), [task = IntrusivePtr<Task>(task)] {
+                 task->Retry();
+               }));
+
+    return absl::OkStatus();
+  }
+
+
+  std::shared_ptr<HttpTransport> transport_;
+  internal_storage_gcs::NoRateLimiter no_rate_limiter_;
+  std::string endpoint_;  // endpoint url
+  SpecData spec_;
+
+  absl::Mutex credential_provider_mutex_;
+  std::optional<std::shared_ptr<CredentialProvider>> credential_provider_;
+  std::string aws_region_;
+};
+
+
+Future<kvstore::DriverPtr> S3KeyValueStoreSpec::DoOpen() const {
+  auto driver = internal::MakeIntrusivePtr<S3KeyValueStore>();
+  driver->spec_ = data_;
+  driver->transport_ = internal_http::GetDefaultHttpTransport();
+
+  if(absl::StripAsciiWhitespace(data_.endpoint).empty()) {
+    // Assume AWS
+    // Make global request to get bucket region from response headers,
+    // then create region specific endpoint
+    auto url = tensorstore::StrCat("https://", data_.bucket, ".s3", kDotAmazonAwsDotCom);
+    auto future = driver->transport_->IssueRequest(
+            HttpRequestBuilder("GET", url).BuildRequest(),
+            absl::Cord());
+    if(!future.status().ok()) return future.status();
+    auto & headers = future.value().headers;
+    if(auto it = headers.find(kAmzBucketRegionHeader); it !=headers.end()) {
+      driver->aws_region_ = it->second;
+      driver->endpoint_ = tensorstore::StrCat(
+        "http://", data_.bucket, ".s3.", driver->aws_region_, kDotAmazonAwsDotCom);
+    } else {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("bucket ", data_.bucket, " does not exist"));
+    }
+  } else {
+    auto parsed = internal::ParseGenericUri(data_.endpoint);
+    if(parsed.scheme != "http" && parsed.scheme != "https") {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("Endpoint ", data_.endpoint,
+                            " has invalid schema ", parsed.scheme,
+                            ". Should be http(s)."));
+    }
+    if(!parsed.query.empty()) {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("Query in endpoint unsupported ", data_.endpoint));
+    }
+    if(!parsed.fragment.empty()) {
+      return absl::InvalidArgumentError(
+        tensorstore::StrCat("Fragment in endpoint unsupported ", data_.endpoint));
+    }
+
+    driver->aws_region_ = "";
```

Comment:
Custom endpoints might still need to set the region. See for e.g.

- https://github.com/minio/minio/issues/14325

I guess an `aws_region` attribute on the driver spec would be sensible.

Reply:
Yes, makes sense.

Reply:
I've added this